### PR TITLE
topology: cleanup/align SoundWire device numbers

### PR DESCRIPTION
**Pull request by plbossart**

### CODE REVIEW

**Review by naveen-manohar — APPROVED** [review]
With UCM changes, Verified Basic Audio on Volteer-SDW  - sof-tgl-sdw-max98373-rt5682.tplg

**Comment by plbossart** [comment]
@bardliao @RanderWang can you take a look on other SoundWire devices?

**Comment by bardliao** [comment]
I tested 3 sdw laptops and all of them looks good to me.
Below is the result of `cat /proc/version`
```
tgl sdca
00-00: Headphone (*) :  : playback 1
00-01: Headset mic (*) :  : capture 1
00-02: SDW1-speakers (*) :  : playback 1
00-04: Microphones (*) :  : capture 1
00-05: HDMI1 (*) :  : playback 1
00-06: HDMI2 (*) :  : playback 1
00-07: HDMI3 (*) :  : playback 1


tgl sdw
00-00: Headphone (*) :  : playback 1
00-01: Headset mic (*) :  : capture 1
00-02: Speakers (*) :  : playback 1
00-04: Microphones (*) :  : capture 1
00-05: HDMI1 (*) :  : playback 1
00-06: HDMI2 (*) :  : playback 1
00-07: HDMI3 (*) :  : playback 1


cml sdw 2 1308 laptop
00-00: Headphone (*) :  : playback 1
00-01: Headset mic (*) :  : capture 1
00-02: SDW1-speakers (*) :  : playback 1
00-04: Microphones (*) :  : capture 1
00-05: HDMI1 (*) :  : playback 1
00-06: HDMI2 (*) :  : playback 1
00-07: HDMI3 (*) :  : playback 1
```

**Comment by RanderWang** [comment]
> I tested 3 sdw laptops and all of them looks good to me.
> Below is the result of `cat /proc/version`
> 
> ```
> tgl sdca
> 00-00: Headphone (*) :  : playback 1
> 00-01: Headset mic (*) :  : capture 1
> 00-02: SDW1-speakers (*) :  : playback 1
> 00-04: Microphones (*) :  : capture 1
> 00-05: HDMI1 (*) :  : playback 1
> 00-06: HDMI2 (*) :  : playback 1
> 00-07: HDMI3 (*) :  : playback 1
> 
> 
> tgl sdw
> 00-00: Headphone (*) :  : playback 1
> 00-01: Headset mic (*) :  : capture 1
> 00-02: Speakers (*) :  : playback 1
> 00-04: Microphones (*) :  : capture 1
> 00-05: HDMI1 (*) :  : playback 1
> 00-06: HDMI2 (*) :  : playback 1
> 00-07: HDMI3 (*) :  : playback 1
> 
> 
> cml sdw 2 1308 laptop
> 00-00: Headphone (*) :  : playback 1
> 00-01: Headset mic (*) :  : capture 1
> 00-02: SDW1-speakers (*) :  : playback 1
> 00-04: Microphones (*) :  : capture 1
> 00-05: HDMI1 (*) :  : playback 1
> 00-06: HDMI2 (*) :  : playback 1
> 00-07: HDMI3 (*) :  : playback 1
> ```

@bardliao  why only 3 HDMI for tgl device ?

File: tools/topology/sof-tgl-rt715-rt711-rt1308-hdmi.m4 [hunk 1]
```diff
@@ -28,9 +28,9 @@ DEBUG_START
 # PCM1 <--- volume <---- ALH 0x103 BE dailink 1
 # PCM2 ---> volume ----> ALH 0x202 BE dailink 2
 # PCM4 <--- volume <---- ALH 0x2 BE dailink 4
-# PCM5 ---> volume <---- iDisp1
-# PCM6 ---> volume <---- iDisp2
-# PCM7 ---> volume <---- iDisp3
+# PCM5 ---> volume ----> iDisp1
+# PCM6 ---> volume ----> iDisp2
+# PCM7 ---> volume ----> iDisp3
```
Comment (RanderWang) [inline tools/topology/sof-tgl-rt715-rt711-rt1308-hdmi.m4:34]: PCM8 for iDisp4 is missed

File: tools/topology/sof-tgl-sdw-max98373-rt5682.m4 [hunk 2]
```diff
@@ -34,12 +34,13 @@ DEBUG_START
...
 # PCM3 <---- demux  <----- ALH0x103 (Speaker -max98373)
-# PCM4 <---- volume <---- DMIC01  (dmic 48k capture)
-# PCM5 <---- kpb <---- DMIC16k  (dmic 16k capture)
-# PCM6 ----> volume -----> iDisp1
-# PCM7 ----> volume -----> iDisp2
-# PCM8 ----> volume -----> iDisp3
+# PCM5 ----> volume -----> iDisp1
+# PCM6 ----> volume -----> iDisp2
+# PCM7 ----> volume -----> iDisp3
 # PCM9 ----> volume -----> iDisp4
```
Comment (RanderWang) [inline tools/topology/sof-tgl-sdw-max98373-rt5682.m4:40]: PCM8 ?

**Comment by RanderWang** [comment]
it is ok on Volteer 
root@sh-tglu-volt-sdw-03:~# cat /proc/asound/pcm
00-00: Headphone (*) :  : playback 1
00-01: Headset mic (*) :  : capture 1
00-02: smart373-spk (*) :  : playback 1
00-03: echo (*) :  : capture 1
00-05: HDMI1 (*) :  : playback 1
00-06: HDMI2 (*) :  : playback 1
00-07: HDMI3 (*) :  : playback 1
00-08: HDMI4 (*) :  : playback 1
00-10: DMIC48kHz (*) :  : capture 1
00-12: DMIC16k (*) :  : capture 1
01-00: USB Audio : USB Audio : playback 1 : capture 1

**Comment by naveen-manohar** [comment]
On Volteer-SDW:
Old_PCM vs New_PCM
image

**Comment by lgirdwood** [comment]
@naveen-manohar are you saying we alos need an UCM update here ?
@plbossart looks like we have a conflict.

**Comment by plbossart** [comment]
> @naveen-manohar are you saying we alos need an UCM update here ?
> @plbossart looks like we have a conflict.

Yes Volteer will need a UCM update, but that was shared and accepted by the Chrome folks. The Volteer configuration is not in alsa-ucm-conf upstream so platforms which are handled upstream, such as Dell/HP, have priority.

**Comment by plbossart** [comment]
> @plbossart looks like we have a conflict.

yes, I managed to self-conflict between PRs. I need to change device names as well.

**Comment by naveen-manohar** [comment]
> > @naveen-manohar are you saying we alos need an UCM update here ?
> > @plbossart looks like we have a conflict.
> 
> Yes Volteer will need a UCM update, but that was shared and accepted by the Chrome folks. The Volteer configuration is not in alsa-ucm-conf upstream so platforms which are handled upstream, such as Dell/HP, have priority.

Post this PR merge, shall take to Chrome Tree & submit along with UCM change.

**Comment by lgirdwood** [comment]
@plbossart fyi - some merge conflicts.

**Comment by plbossart** [comment]
> @plbossart fyi - some merge conflicts.

yes, will work on this later today.

**Comment by plbossart** [comment]
@ranj063 @kv2019i @bardliao FYI here's the list of devices and names I came up with. 

[PCMDeviceList.txt](https://github.com/thesofproject/sof/files/5954259/PCMDeviceList.txt)

**Review by lgirdwood — APPROVED** [review]

**Comment by plbossart** [comment]
@lgirdwood we should have been more careful on this one, CI tests did show a problem with alsa-bat precisely due to the device number issues.

https://sof-ci.01.org/sofpr/PR3806/build8108/devicetest/

I only realized this with the daily test and @fredoh9 's help.

good old alsa-bat still has its value. All other tests don't seem to check for expected device numbers or names.

**Comment by lgirdwood** [comment]
@plbossart thanks, fwiw, I've been seeing this all week and I've created an issue for CI test to dump PCM and card names when this happens. Do you have a followup for UCM ? or do you want to revert and hold for v1.8 ?

**Comment by plbossart** [comment]
> @plbossart thanks, fwiw, I've been seeing this all week and I've created an issue for CI test to dump PCM and card names when this happens. Do you have a followup for UCM ? or do you want to revert and hold for v1.8 ?

I have a followup coming, @ranj063 helped me find the issues in current topologies that were exposed by my changes.
It's literally a Jenga game, when you remove or redefine a part the whole thing can crumble.

**Pull request merged**